### ISSUE TITLE
Improving handling of drives with bad sectors and filesystem inconsistencies

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#237]
One of the [biggest](https://www.opennet.ru/openforum/vsluhforumID3/124448.html#3) criticisms of Clonezilla (and therefore Rescuezilla) is it doesn't handle disks with bad sectors that well:

> The "clonzilla" has a limitation that infuriates at the right moments, whether this thing can read sections with physically damaged cells from ssd, without interrupting copying.

I have also come across a [user](https://sourceforge.net/p/redobackup/discussion/general/thread/a8e8f2570b/) suggesting another partclone-based tool handles this better:

> I'm a fan of of RedoRescue - it works when all other backup and recovery tools can't handle with bad sectors.

This piqued my interest. I know from my implementation of [support for Redo Rescue images in Rescuezilla](https://github.com/rescuezilla/rescuezilla/issues/194) that it's simply just another frontend to `partclone` and a bunch of other tools which the much more popular industry-standard Clonezilla (that Rescuezilla is based on) has been using for decades. Partclone offers two options for handling disk read errors:

> -R,  --rescue           Continue clone while disk read errors
> -F,  --force            Force progress

The above description of partclone's "--rescue" option is simple enough, but for "--force", the partclone source code is more descriptive, suggesting it is used to "skip the metadata error" when the "Used Block count is different".

I tried searching the Redo Rescue code, and the program appears to use the `--force` option, but _not_ `--rescue`:
```
redorescue$ git grep -- "--rescue\| -R\|--force\| -F"
overlay/rootdir/var/www/html/functions.inc.php: $cmd = "partclone.$fs_tool $fs_mode --force --UI-fresh 1 --logfile ".TMP_DIR."$dev.log ";
overlay/rootdir/var/www/html/functions.inc.php: $cmd .= " | partclone.$fs_tool --restore --force --UI-fresh 1 ";
overlay/rootdir/var/www/html/functions.inc.php: $cmd .= " | partclone.chkimg --force --UI-fresh 1 --source - ";
```
I am **extremely hesitant** to add "--force" as a **default** Rescuezilla partclone option because the metadata check around the used block count is the first defense to ensure consistency of the filesystem. Using "--force" **introduces the very risky situation of filesystem inconsistencies propagating into the backup images** and where the user doesn't realize their filesystem (or drive itself) is in a bad state, and therefore continues to use the drive. This could very easily lead to data loss, especially since partclone is not designed for damaged filesystems. Running a simple filesystem check before backup (like Clonezilla offers to) may help, but definitely doesn't fully address this risk.

I think the approach Clonezilla (and Rescuezilla) takes by failing to create such backups by default is the right approach. It informs the user that their drive is unhealthy (whether it's a filesystem issue or a bad sectors), and they can take measures to address this. If a filesystem check is insufficient, a user can still backup/clone using Clonezilla's sector-by-sector 'dd' tool rather than partclone, or in Rescuezilla's case the future [ddrescue-based sector-by-sector "data salvage" mode (task #143)](https://github.com/rescuezilla/rescuezilla/issues/143). But sector-by-sector backups are large and slow, so this is not that great.

Clonezilla has a "-rescue" option, but from the source code it seems to only enable partclone's "--rescue" option, and not partclone's "--force" option. I suggest that Clonezilla's "-rescue" should enable both.

<a href="https://clonezilla.org/clonezilla-live/doc/03_Disk_to_disk_clone/images/ocs-05-b-2-expert-param-extra.png"><img width=600 src="https://clonezilla.org/clonezilla-live/doc/03_Disk_to_disk_clone/images/ocs-05-b-2-expert-param-extra.png"></a>

As for Rescuezilla, it makes sense to add some kind of advanced option to the Rescuezilla user-interface to allow advanced users to enable partclone's "--rescue" and "--force" options during backup and clone operations. If there's no clean way to add the option, I might have to add an 'advanced options' page to the wizard (something I've been desperately trying to avoid). Maybe a page where you can select between partclone, ntfsclone etc (which is implemented by Clonezilla's Expert Mode and other tools like FOG Project, but not Rescuezilla) and configure some advanced tool-specific options.

EDIT: On further thought, enabling `--force` and `--rescue` during _clone_ operations should happen by default. This allows people with damaged source drives to move to a healthy disk without the previously mentioned risk of silently propagating filesystem inconsistencies propagating into an image. But also exposing a user-interface option for backup/restore/clone modes is worthwhile.

Side note: I don't yet think it makes sense to integrate any partclone options into the ["data salvage" mode (task #143)](https://github.com/rescuezilla/rescuezilla/issues/143), the focus of the "data salvage" mode will be on ddrescue.

**Comment by shasheene** [comment]
> Clonezilla has a "--rescue" option, but from the source code it seems to only enable partclone's "--rescue" option, and not partclone's "--force" option. I suggest that Clonezilla's "--rescue" should enable both.

@stevenshiau, what do you think of the above statement?

From my reading of the Clonezilla source code it looks like there's not yet an option to call partclone with "--force", to skip checking the used-block count metadata but still make a filesystem-aware backup. Partclone exposes this option, so Clonezilla ought to too. But adding it to Clonezilla's existing "--rescue" option seems like the nicest solution?

**Comment by stevenshiau** [comment]
Sure. I will append "--force" to the existing "--resuce" option. Thanks for this suggestion.

**Comment by shasheene** [comment]
**And what about adding a `-rescue` option for Clonezilla's `restoredisk`, `restoreparts`?**

As best I can tell, Clonezilla's Expert Mode currently does **NOT** provide this option for restore operations. Only for backup (savedisk/saveparts) and cloning. But it would definitely help users if the option was available!

**Comment by stevenshiau** [comment]
Oh, actually the '-rescue' is accepted in the command line option, but it's not added in the TUI. It's a bug. We should add that. Thanks. It will be added in the next release.

**Comment by shasheene** [comment]
Addressed in the upcoming Rescuezilla v2.3 release with a user-interface element on the confirmation screen:

![rescue_option](https://user-images.githubusercontent.com/1414379/142799098-8826bc7a-7677-4a26-8dc3-07cce6d3b96d.png)

As more advanced options are added, the toggle checkbox UI element will probably be moved to an "advanced options" page like Clonezilla itself, rather than cluttering the confirmation page.
